### PR TITLE
Update session metadata

### PR DESCRIPTION
**Pull request by XX-Yin**
### Pull Request instructions:
- Please follow the [update protocol](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki/Software-Update-Procedures)
- Answer the questions below in detail. Your responses will be emailed to experimenters. 
- If the experimenters must do anything new, provide detailed step by step instructions on the [wiki](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki)
- If computer maintainers need to manually update anything, provide detailed step by step instructions
- Use markdown syntax in order for your comments to be rendered reliably in the email: "1." instead of "1)", use four spaces for indents.
- If you use the keyword "skip email" in the title, it will skip the email updates
- Merges from "develop" into "production_testing" should use the keyword "production merge" in the title for reliable indexing of updates
- Merges from "production_testing" into "main" should use the keyword "update main"
  
### Describe changes:
1. Updated the session metadata generation to handle a lot of edge cases we found. 
2. When loading a session into the GUI and clicking the "Save" button, the metadata will be regenerated using the existing data saved in the behavior json. The behavior json will not be resaved.
3. Previously logging with `logging.error(str(e))` would not log the full error information, change it to `logging.error(traceback.format_exc())` to get the full traceback error.
4. Enabling metadata to be generated directly from the behavior json (ideally, if the metadata dialog is filled in, the behavior json contains all the information to generate metadata).
5. Separating the behavior running version and metadata generation software version, which enables us to track the metadata generation version as the metadata may be updated after behavior running (for example, adding more behavior metrics). 
6. Adding metadata generation logging when `GenerateMetadata.py` is used independently. The default logging folder is `C:\Users\user_name\Documents\foraging_metadata_logs`. 
7. Adding logging when return is triggered. 

### What issues or discussions does this update address?
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/573 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/569 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/525 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/510 partially resolved
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/494 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/493 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/491 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/490 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/469 resolved 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/468 not resolved in the PR. Waiting for careful sound calibration. 
https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/527#issuecomment-2177172152 Temporarily resolved. Output parameters (including task parameters and performance metrics) are temporarily saved in the `output_parameters` in the auditory `StimulusEpoch`, which does not make sense as these parameters are for that session instead of a specific stimulus. I have submitted an issue to computation team to ask for changes https://github.com/AllenNeuralDynamics/aind-data-schema/issues/993. 

### Describe the expected change in behavior from the perspective of the experimenter
You can regenerate the metadata for old sessions that have the metadata dialog filled in by running `generate_metadata(json_file=behavior json)`. 

### Describe any manual update steps for task computers
No

### Was this update tested in 446/447?
Tested on my own computer with old sessions from 447 and 323_EPHYS3

### CODE REVIEW

**Comment by XX-Yin** [comment]
I have tested generating session metadata in `323_EPHYS3` and `447 behavior rigs` for sessions that have the metadata dialog. 

`323_EPHYS3`, `9mice`, `118 sessions`, `all successful`

`447 behavior rigs`, `33 mice`, `546 sessions`, `541 successful`, `5 error` (shown below). 

All error sessions are related to missing some behavior data. This is most likely due to the RA loading the session and hitting the save button, causing the data to be lost. I have fixed this issue in this PR. 

Error generating metadata for Z:\svc_aind_behavior_transfer\447-1-D\700708\behavior_700708_2024-06-13_08-53-39\behavior\700708_2024-06-13_08-53-39.json: 'BS_TotalReward'
Error generating metadata for Z:\svc_aind_behavior_transfer\447-2-D\713310\behavior_713310_2024-06-11_08-49-17\behavior\713310_2024-06-11_08-49-17.json: 'BS_TotalReward'
Error generating metadata for Z:\svc_aind_behavior_transfer\447-2-D\714728\behavior_714728_2024-07-02_10-01-11\behavior\714728_2024-07-02_10-01-11.json: 'BS_TotalReward'
Error generating metadata for Z:\svc_aind_behavior_transfer\447-2-D\716868\behavior_716868_2024-06-06_16-32-52\behavior\716868_2024-06-06_16-32-52.json: 'BS_TotalReward'
Error generating metadata for Z:\svc_aind_behavior_transfer\447-3-D\712634\behavior_712634_2024-06-04_10-58-36\behavior\712634_2024-06-04_10-58-36.json: 'BS_TotalReward'

**Comment by XX-Yin** [comment]
@hagikent @alexpiet @hanhou @ZhixiaoSu @KanghoonJ

This PR is ready for review!

**Comment by hagikent** [comment]
@XX-Yin Great job! Thanks for your effort! I will test this over the weekend.

**Comment by hagikent** [comment]
@XX-Yin 
Just a quick clarification: I saw two different day-time formats intermingled in the previous code and causing errors at times. I presume you decided to stick to `yyyy-mm-dd_hh-mm-ss`, correct?

**Comment by XX-Yin** [comment]
> @XX-Yin Just a quick clarification: I saw two different day-time formats intermingled in the previous code and causing errors at times. I presume you decided to stick to `yyyy-mm-dd_hh-mm-ss`, correct?

You mean for some old sessions? I think they are all unified now as `%Y-%m-%d_%H-%M-%S`. 

Can you send me the date format for these error sessions?

File: src/foraging_gui/Foraging.py [hunk 1]
```diff
@@ -2517,31 +2527,14 @@ def _GetSaveFolder(self):
             photometry data
             ephys data
         '''
-        current_time = datetime.now()
-        formatted_datetime = current_time.strftime("%Y-%m-%d_%H-%M-%S")
-        self.session_name = f'behavior_{self.ID.text()}_{formatted_datetime}'
-        self.acquisition_datetime = current_time.strftime("%Y-%m-%d %H:%M:%S")
```
Comment (hagikent) [inline src/foraging_gui/Foraging.py:2523]: here for example. but i think all fixed now!

**Review by hagikent — COMMENTED** [review]
You had several  `%Y-%m-%d %H:%M:%S` but seemingly all fixed now.

**Review by alexpiet — APPROVED** [review]

**Pull request merged**